Add introductory documentations (#77)
postech-dao/simperby: docs/protocol_overview.md

# Simperby: Protocol Overview

Simperby is a blockchain engine that runs an organiziation which is

1. **Decentralized**
2. **Standalone, Sovereign and Self-hosted**
3. **With a set of permissioned members**

in that it provides

1. **Consensus**: a safe and live mechanism to finalize the state of the organization, which is tolerant to byzantine faults in an asynchronous network.
2. **Governance**: a democratic mechanism to make decisions on the organization, which can be finalized by the consensus.
3. **Communication Channel**: a mechanism to communicate with each other in a decentralized, verifiable, and fault-tolerant way.

## Basics and Terminologies

The reader of this document should be familiar with the basic concepts of blockchain.

- Simperby is an engine that can build an instance of a blockchain.
  Any blockchain built with Simperby is sometimes called a *Simperby chain*.
- The unit of state transition is called a *transaction*.
  `StoreData`, `DelegateGovernanceVotingPower`, `AddValidator` are the examples.
- A sequence of transactions is called an *agenda*.
  **Any governance voting is performed on an agenda and the target block height**.

> - A *block* consists of the header and **a single agenda, which is mandatory**.

Author: updated

- The participants of the governance are called *members*.
- The participants of the consensus are called *validators*, which is a subset of the *members*.
- A *consensus leader* or a *block proposer* is a validator that proposes a block for the round.
- A *round* is a period of time in which a consensus leader proposes a block, and consensus votes(pre-vote, pre-commit) are performed. Refer to Tendermint for more details.
- A validator is *honest* or *not byzantine* if they follow the consensus protocol.
- A block proposer is *faithful* if they fulfill their responsibility and *good* if they don't overuse their power. They are *lazy* and *bad* otherwise.
  The notion of *responsibility* and *power* will be explained [later](#consensus-leader).

## Summary

Here is a summarized version of the Simperby protocol.

1. We want the node operation to be **simple and lightweight** to realize a truly distributed, decentralized and self-hosted organization.

> 2. To be so, it is assumed that most of the nodes are **rarely online** and the protocol produces new blocks **on-demand**.

Reviewer: 1. The meaning of "To be so" is ambiguous. (Guess you mean "organization"?)
2. The sentence can be shortened.
```suggestion
3. In such an organization, most of the nodes are assumed **rarely online** and the protocol produces new blocks **on-demand**.
```

Author: 1. To be 'a truly distributed, decentralized and self-hosted organization.'
2. I'll change to 'the key assumption is'

3. To accomplish that, the **consensus round must be very long**.

> 4. For every block, members can **vote on or propose an agenda, propagating** to each other.

Reviewer: Propagating what?
```suggestion
4. For every block, members can **vote on or propose an agenda**. Votes and agendas are propagated to every node.
```

Author: 'to each other' sounds more gossip-protocol-like.

Reviewer: I mean there's no subject of "propagating".

Author: How about `For every block, members can **vote on or propose an agenda, propagated to each other**

Author: okay.

> 4. For every block, members can **vote on or propose an agenda**. Agendas and votes are **propagated to each other**.

Author: updated

> 5. At the same time, members can also **propagate their chats** used for the communication, which is **coordinated by the block proposer**.

Reviewer: I think the word "coordinate" is not that appropriate. How about using "gather" or "collect" instead?

Author: The proposer doesn't just collect the chats. It frequently semifinalize the chat chain, and that's the point.

Reviewer: I see

> 5. At the same time, members can also **propagate their chats** used for the communication, which is **coordinated by the block proposer**.

Author: updated

> 6. If there is a **governance-approved (majority-voted) agenda** on the network, the block proposer should **include it in the block along with the chat log**, and propose to the consensus.

Author: The threshold MUST be higher than 1/2 anyway. (remember that we don't collect votes against the agenda)
Do we still need to mention that the threshold may go higher than 1/2 if the org wants?

7. In that the block proposer should be online most of the time (chat coordination and block proposing), this **responsibility should be laid on a few of the validators most of the time**, to ensure the rarely-online assumption.
8. Also the role of **block proposer has some authorities** (chat coordination and agenda inclusion) that can't be cryptographically verified if misused (typically censorship).

> 9. Thus there exists **'few of the validators' that can be lazy or commit harmful misuses** of their authorities which aren't really byzantine faults or invalid blocks.

Reviewer: ```suggestion
9. Thus there exists **'few validators' that can be lazy or commit harmful misuses** of their authorities which aren't really byzantine faults like proposing invalid blocks.
```

Author: The meaning has beeen changed.
Mine: there is a privileged set of validators that must be moderated.
Yours: it is natural that some validators may become be lazy or harmful.

Author: Also, I distinguished 'byzantin fault' and 'invalid block' but you included the latter in the former.

Reviewer: > Also, I distinguished 'byzantin fault' and 'invalid block' but you included the latter in the former.

ok I understand

Reviewer: > Mine: there is a privileged set of validators that must be moderated.

Well, I'm not sure the readers will comprehend what you mean with the original sentence.
I'll resolve this conversation if other reviewers are happy with this.

Author: I changed it to a super explicit expression.

Reviewer: great

> 10. Normally this isn't a problem, if the rounds are short and the block proposer changes regularly, but we're not.

Reviewer: ```suggestion
10. This isn't a problem if the rounds are short and the block proposer is changed regularly, unlike in our case.
```

Author: It is important to emphasize 'normally'.

11. Therefore, we need a special mechanism to **'veto' the block proposer not to waste long rounds in the consensus layer**. Thus we introduce a special variation of Tendermint called *Vetomint*.

> 12. In Vetomint, validators can veto the current block proposer, but still, the round will progress (changing the block proposer) if all the honest validators either vote or veto.

Author: updated

## Simple and Light Node Operation

Simperby strongly encourages each member to run their node,
to ensure that the network is decentralized and distributed (and so truly self-hosted). In other words, running a Simperby chain is not about using AWS servers by a few people but using their laptops to physically run the chain. To achieve that, it is important to **keep the node operation as simple and lightweight as possible**.

### Rarely-online nodes

One of the most important conditions for 'simple' and 'lightweight' node operations is how often the node needs to be online.
If we want to make the validators run nodes on their laptops, it is not realistic to assume that they will be online 24/7.

Since Simperby's BFT consensus assumes a partially-synchronous network, rarely-online nodes can be trivially handled because it's no more than one typical case of an asynchrony, if

1. the consensus round is (or has grown to be) long enough to cover all appearances of the nodes.

> 2. at least one node is online when there is a network broadcast

Author: updated

It's super-easy to expand the organization to other existing chains (so-called 'colony chains') using
the light client of the Simperby consensus.
Since the light client is uploaded as a contract to other chains,
any block progress of the Simperby consensus will require an additional
transaction to update and store the Merkle root of it.
This might not be cheap, because **the organization will have to pay for the gas cost for every colony chain, for every Simperby block**.
That is another reason that on-demand block production is considered reasonable.

## Governance

Simperby's governance is implemented by P2P voting.

- Any member can propose an agenda, and it will be propagated over the P2P network.
- Any member also can cast a vote on the agenda if they prefer it. The vote will be propagated as well.
- Any agenda that has over 50% votes is **eligible** to be included in a block.
- A block can be **valid only if it includes an eligible agenda**. In other words, empty blocks are not allowed.
- To decide which eligible agenda to finalize is the role of consensus, not the governance.
- The consensus is a BFT algorithm based on a leader-and-round style.

> ### What can be an agenda?

Reviewer: How about adding a description of the voting process in this section?
(a sentence like "A transaction includes a human-readable description of it, and each individual reads the description to determine whether to cast a vote or not.")

Author: Actually, it's not that simple. (especially after I introduce the transaction-as-a-git-commit protocol)
The definitions of 'transaction', 'state', and 'state transition' are beyond the scope of this documents.
Focusing only on the governance and consensus, leaving the higher layer abstract will make this document more concise.

Reviewer: Okay. I agree with you.

- Transaction is only the unit of a **state transition**, not an individual item signed and broadcasted as in the case of a conventional blockchain.
- An 'agenda' is defined as `(Proposer, Height, [Transaction])`.
- Only a **single agenda** can be included in a block. This is for preventing complicated dependency problems.
  - If multiple agendas are allowed, each agenda must be independent; Otherwise, voters can't be sure how each agenda is ordered, or even whether it is included in the finalized block. This will cause a severe restriction of the possible agenda items.
- Because `Height` is a part of an agenda,

>   every vote will be outdated and thus discarded if the block height progresses.

Reviewer: Because `Height` is a part of an agenda, every vote will be outdated and thus discarded if the block height progresses.
Do you mean "agenda"?  ------------------------^^

Author: Both agenda and vote. I will update.

Reviewer: ok

>   every agenda and its vote will be outdated and thus discarded if the block height progresses.

Author: fixed

- Nonetheless, if the leader attempts censorship more, it becomes more suspicious.

Note that if there is a malicious member who tries to spam the chat, the leader will not finalize the spam chain, and the other members would recognize such censorship as something `not bad`.

### What happens in the consensus?

The consensus leader (*block proposer*; don't be confused with the *agenda proposer*) includes an eligible agenda (if there were multiple eligible agendas, the leader just chooses one) in the block, and proposes it.
**A valid block MUST contain an eligible agenda.**. In other words, if there is no eligible agenda, there is no block progress in the blockchain.

In addition to the agenda, the leader can also include some other transactions.

1. RecordChatLog (chatting logs for the very height): this doesn't change the state, but is used as the governance minutes for the agenda.

> 2. ReportMisbehavior (double votes in the consensus and chat finalization misbehaviors by a past consensus leader): this will lead to the slashing of the voting power of the misbehaving validator.

Reviewer: Then how the voting power be slashed? Is it automated, or does it require proposing an agenda?

Author: Since it's not the part of an agenda, it does not require a governance approval.

Author: This is a common design in almost every stake-based chains, so looks quite clear to me. I will put a word 'automatically' just in case.

Reviewer: okay

> 2. ReportMisbehavior (double votes in the consensus and chat finalization misbehaviors by a past consensus leader): this will automatically lead to the slashing of the voting power of the misbehaving validator.

Reviewer 2: How do we implement slashing?

Author: getting rid of the validator out of the set.

Another advantage of a stable leader is that it is **predictable**. Every validator can be quite sure how often they should turn on their node based on the order in the stable leader list.

As long as the selected first few leaders are honest(not byzantine), and *faithful* (fulfilling their responsibilities), the network will be **live**.

### Authority

The leader has **authority**.

- The leader can choose which agenda to include in the block if there are multiple eligible agendas.
- The leader can choose which chat chain to semifinalize.

A *good* node should be objective to agendas; they should choose the first eligible agenda, instead of waiting until it observes another one in their personal favor. Similarly, a 'good' node should be objective to the chat chain semifinalization as well.

As long as the selected first few leaders are honest, faithful, and even good,
the network will be not only **live** but also **fair**. (note that it's always **safe** regardless of the leader's behavior; safety is guaranteed by the BFT consensus)

> ### So What?

Reviewer: Let's use more meaningful words for the title instead of context-depending words.
(e.g. dealing with a lazy or bad leader)

Author: updated

Since Tendermint is based on rounds, there is a special mechanism called *nil-vote*. Nodes can cast a nil-vote after the timeout, to escape the current round. This works similarly to the non-nil case, in that it takes two steps.

In plain Tendermint, if the validators use nil-votes to veto the proposer, it just works because it's not so different from a rash timeout, but it might get stuck until the timeout in the split case. (Think about the case of 6/10 non-nil-votes and 3/10 nil-votes, 1/10 Unresponsive Byzantine nodes)
Again, because we're using very long rounds, it will waste a lot of time, waiting for the timeout of the round in that kind of situation. If we encounter this unlucky case of a split, it is no better than the governance's solution (no approval of agendas).

### Vetomint

To improve this, we propose a variation of Tendermint, called *Vetomint*, which is effective in the 'split' case.

### Properties

1. `<1/6` BFT
2. It is allowed (i.e., considered as a non-byzantine behavior) to cast a nil-vote before the timeout, with *some kind of reason*.

> 3. If all the honest nodes cast either a vote or nil-vote, it is guaranteed that the round progresses; the **early termination**.

Reviewer: In this sentence and below, how about consistently using "non-nil-vote" for some clarity?

Author: updated

4. If all the honest nodes cast a vote, it is guaranteed that the block progresses.
5. If over 2/3 of the nodes cast a vote, the block progresses with the probability of `>0`.

### How does it work?

In Vetomint, unlike Tendermint, the round **always instantly progresses** if over 5/6 nodes either non-nil-vote or nil-vote. Especially when 5/6 nodes vote, it is guaranteed that the height progresses. The threshold 5/6 is just enough to ensure that all nodes observe at least 2/3 (Tendermint majority) of non-nil-votes no matter the individual arrival order of the votes, even with the possible 1/6 nil-votes.
Note that the 'arrival order' matters, unlike Tendermint, because we have a new 'early termination' condition.

> Details are explained in issue #4 and the pseudo-code is [here](./vetomint-spec.pdf)

Reviewer: Add a link to the issue.
https://github.com/postech-dao/simperby/issues/4

Author: updated
